subt/docker/robotika/ros_proxy_node.cc

```diff
@@ -0,0 +1,531 @@
+                                   const uint32_t _dstPort,
+                                   const std::string &_data);
+
+  /// \brief ROS node handler.
+  private: ros::NodeHandle n;
+
+  /// \brief publisher to send cmd_vel
+  public: ros::Publisher velPub;
+
+  /// \brief Communication client.
+  private: std::unique_ptr<subt::CommsClient> client;
+
+  /// \brief Client to request pose from origin.
+  ros::ServiceClient originClient;
+
+  /// \brief Service to request pose from origin.
+  subt_msgs::PoseFromArtifact originSrv;
+
+  ros::Subscriber subClock;
+  ros::Subscriber subImu;
+  ros::Subscriber subScan;
+  ros::Subscriber subImage;
+  ros::Subscriber subOdom;
+
+  /// \brief True if robot has arrived at destination.
+  public: bool arrived{false};
+
+  /// \brief True if started.
+  private: bool started{false};
+
+  /// \brief Last time a comms message to another robot was sent.
+  private: std::chrono::time_point<std::chrono::system_clock> lastMsgSentTime;
+
+  /// \brief Name of this robot.
+  private: std::string name;
+
+  private: double prev_dist2{0.0};
+
+  public: bool ReportArtifact(subt::msgs::Artifact& artifact)
+  {
+    std::string serializedData;
+    if (!artifact.SerializeToString(&serializedData))
+    {
+      ROS_ERROR("ReportArtifact(): Error serializing message [%s]",
+          artifact.DebugString().c_str());
+      return false;
+    }
+    else if (!this->client->SendTo(serializedData, subt::kBaseStationName))
+    {
+      ROS_ERROR("CommsClient failed to Send serialized data.");
+      return false;
+    }
+    return true;
+  }
+
+  public: bool getSpeedCmd(geometry_msgs::Twist& msg);
+  public: bool parseArtf(char *input_str, subt::msgs::Artifact& artifact);
+};
+
+/////////////////////////////////////////////////
+Controller::Controller(const std::string &_name)
+{
+  ROS_INFO("Waiting for /clock, /subt/start, and /subt/pose_from_artifact");
+
+  ros::topic::waitForMessage<rosgraph_msgs::Clock>("/clock", this->n);
+
+  // Wait for the start service to be ready.
+  ros::service::waitForService("/subt/start", -1);
+  ros::service::waitForService("/subt/pose_from_artifact_origin", -1);
+  this->name = _name;
+  ROS_INFO("Using robot name[%s]\n", this->name.c_str());
+}
+
+/////////////////////////////////////////////////
+void Controller::CommClientCallback(const std::string &_srcAddress,
+                                    const std::string &_dstAddress,
+                                    const uint32_t _dstPort,
+                                    const std::string &_data)
+{
+  subt::msgs::ArtifactScore res;
+  if (!res.ParseFromString(_data))
+  {
+    ROS_INFO("Message Contents[%s]", _data.c_str());
+  }
+
+  // Add code to handle communication callbacks.
+  ROS_INFO("Message from [%s] to [%s] on port [%u]:\n [%s]", _srcAddress.c_str(),
+      _dstAddress.c_str(), _dstPort, res.DebugString().c_str());
+}
+
+/////////////////////////////////////////////////
+void Controller::Update()
```

Comment:
This function is somewhat long. I would most likely do the initialization outside of `Update` and called it explicitly before entering the main loop.